cmd/kubeadm/app/phases/upgrade/health.go

```diff
@@ -179,35 +186,37 @@ func createJob(client clientset.Interface, cfg *kubeadmapi.ClusterConfiguration)
 		return false, nil
 	})
 	if err != nil {
-		return errors.Wrapf(lastError, "Job %q in the namespace %q did not complete in %v", jobName, ns, timeout)
+		// Return the error as a warning for this preflight check
+		return errors.Wrapf(lastError, "Job %q in the namespace %q did not complete in %v. "+
+			"NOTE: This is expected to fail in single node clusters if the node is cordoned prior to upgrade", jobName, ns, timeout), nil
 	}
 
 	klog.V(2).Infof("Job %q in the namespace %q completed", jobName, ns)
 
-	return nil
+	return nil, nil
 }
 
 // controlPlaneNodesReady checks whether all control-plane Nodes in the cluster are in the Running state
-func controlPlaneNodesReady(client clientset.Interface, _ *kubeadmapi.ClusterConfiguration) error {
+func controlPlaneNodesReady(client clientset.Interface, _ *kubeadmapi.ClusterConfiguration) (error, error) {
 	selectorControlPlane := labels.SelectorFromSet(map[string]string{
 		constants.LabelNodeRoleControlPlane: "",
 	})
 	nodes, err := client.CoreV1().Nodes().List(context.TODO(), metav1.ListOptions{
 		LabelSelector: selectorControlPlane.String(),
```

Comment:
we already list nods in another place but this is with a label selector, so i would assume it's much faster than:
- listing all of them during the CreateJob preflight
- don't create the job if "nodes == 1"

?

this is the alternative way of fixing this @SataQiu @pacoxu @carlory 
if this list of nodes here is already slow, then maybe we can list all nodes in one place before running the preflight checks.

Reply:
how about this?

IF `kubectl get no --field-selector spec.unschedulable=false --chunk-size=1` is not empty:
    do CreateJob
ELSE
    print a warning message

Reply:
> I'm not sure if it is fast or slow to list all the nodes or parts in the cluster.

that's the Q here really if filtering the nodes with node selectors and label selectors is MUCH faster then we can just check if there is a single CP node or 

> how about this?
> 
> IF `kubectl get no --field-selector spec.unschedulable=false --chunk-size=1` is not empty: do CreateJob ELSE print a warning message

i like the idea. i will update the PR later.

Reply:
@carlory updated, PTAL